src/data/setoid.lean

```diff
@@ -0,0 +1,487 @@
+  injective f ↔ ker f = ⊥ :=
+⟨λ hf, setoid.ext' $ λ x y, ⟨λ h, hf h, λ h, h ▸ rfl⟩,
+ λ hk x y h, show rel ⊥ x y, from hk ▸ (show (ker f).rel x y, from h)⟩
+
+/-- The elements related to x ∈ α by the kernel of f are those in the preimage of f(x) under f. -/
+lemma ker_apply_eq_preimage (f : α → β) (x) : (ker f).rel x = f ⁻¹' {f x} :=
+set.ext $ λ x,
+  ⟨λ h, set.mem_preimage.2 (set.mem_singleton_iff.2 h.symm),
+   λ h, (set.mem_singleton_iff.1 (set.mem_preimage.1 h)).symm⟩
+
+/-- The uniqueness part of the universal property for quotients of an arbitrary type. -/
+theorem lift_unique {r : setoid α} {f : α → β} (H : r ≤ ker f) (g : quotient r → β)
+  (Hg : f = g ∘ quotient.mk) : quotient.lift f H = g :=
+begin
+  ext,
+  rcases x,
+  erw [quotient.lift_beta f H, Hg],
+  refl
+end
+
+/-- Given a map f from α to β, the natural map from the quotient of α by the kernel of f is
+    injective. -/
+lemma injective_ker_lift (f : α → β) : injective (@quotient.lift _ _ (ker f) f (λ _ _ h, h)) :=
+λ x y, quotient.induction_on₂' x y $ λ a b h, quotient.sound' h
+
+/-- Given a map f from α to β, the kernel of f is the unique equivalence relation on α whose
+    induced map from the quotient of α to β is injective. -/
+lemma ker_eq_lift_of_injective {r : setoid α} (f : α → β) (H : ∀ x y, r.rel x y → f x = f y)
+  (h : injective (quotient.lift f H)) : ker f = r :=
+le_antisymm
+  (λ x y hk, quotient.exact $ h $ show quotient.lift f H ⟦x⟧ = quotient.lift f H ⟦y⟧, from hk)
+  H
+
+variables (r : setoid α) (f : α → β)
+
+/-- The first isomorphism theorem for sets: the quotient of α by the kernel of a function f
+    bijects with f's image. -/
+noncomputable def quotient_ker_equiv_range :
+  quotient (ker f) ≃ set.range f :=
+@equiv.of_bijective _ (set.range f) (@quotient.lift _ (set.range f) (ker f)
+  (λ x, ⟨f x, set.mem_range_self x⟩) $ λ _ _ h, subtype.eq' h)
+  ⟨λ x y h, injective_ker_lift f $ by rcases x; rcases y; injections,
+   λ ⟨w, z, hz⟩, ⟨@quotient.mk _ (ker f) z, by rw quotient.lift_beta; exact subtype.ext.2 hz⟩⟩
+
+/-- The quotient of α by the kernel of a surjective function f bijects with f's codomain. -/
+noncomputable def quotient_ker_equiv_of_surjective (hf : surjective f) :
+  quotient (ker f) ≃ β :=
+@equiv.of_bijective _ _ (@quotient.lift _ _ (ker f) f (λ _ _, id))
+  ⟨injective_ker_lift f, λ y, exists.elim (hf y) $ λ w hw, ⟨quotient.mk' w, hw⟩⟩
+
+/-- The third isomorphism theorem for sets. -/
+noncomputable def quotient_quotient_equiv_quotient (s : setoid α) (h : r ≤ s) :
+  quotient (ker (quot.map_right h)) ≃ quotient s :=
+quotient_ker_equiv_of_surjective _ $ λ x, by rcases x; exact ⟨quotient.mk' x, rfl⟩
+
+/-- Given an equivalence relation r on α and a surjective map f from α to the quotient of α by r,
+    the natural map from equivalence relations containing r to equivalence relations on the 
+    quotient of α by r. -/
+def to_quotient {f : α → quotient r} (hf : surjective f) (s : {s // r ≤ s}) :
```

Comment:
I think I wasn't clear here. I don't think the type should be `quotient r`. Sometimes there is a type that is morally a quotient, but is not actually constructed as a quotient (integers mod 2 are a quotient of integers mod 4). If there is surjective map from X to Y, then Y is isomorphic to some quotient of X - that's what I meant by generalising to surjective maps. I think lemmas about the quotient should usually be stated in this generality.

Reply:
Sorry, I was away from Lean today and I should be able to deduce this but do you mean keeping things the same except for ```f : α  → β``` and mapping to ```setoid β``` or something like
```lean
def aux {f : α → β} (hf : surjective f) (s : {s // ker f ≤ s}) :
  setoid (quotient (ker f)) := sorry

def to_quotient {f : α → β} (hf : surjective f) (s : {s // ker f ≤ s}) :
  setoid β :=
let q := (quotient_ker_equiv_of_surjective f hf).symm in
⟨λ x y, (aux hf s).rel (q x) (q y), sorry⟩
```
more in keeping with the 'morally a quotient' comment? At the moment the latter's giving me the error "failed to generate bytecode for 'to_quotient''
code generation failed, VM does not have code for 'classical.indefinite_description'"

Reply:
That looks good. If there isn't a concrete reason to do use subtypes I would probably unbundle `(s : {s // ker f ≤ s})`.

You should be able to define the relation constructively. It would be something like `λ x y, ∃ a b, s.rel a b ∧ f a = x ∧ f b = y `. Thinking about it now it might be better to get rid of the surjectivity assumption, rename it `map` and define it as the equivalence closure of the image. This is probably functorial. There might be certain nicer properties in the case of surjectivity. You could also rename `of_quotient` to `comap`